core/storage/statistics/gae_models.py

```diff
@@ -1134,6 +1138,132 @@ def delete_playthroughs_multi(cls, playthrough_ids):
+        [ENTITY_TYPE].[ENTITY_ID].[GENERATED_STRING]
+    """
+    # ID of the entity.
+    entity_id = ndb.StringProperty(required=True, indexed=True)
+    # The type of entity, that whether it is exploration, question, or any
+    # other object if it is added in future.
+    entity_type = ndb.StringProperty(required=True, indexed=True)
+    # The answer submitted by the learner.
+    learner_answer = ndb.StringProperty(required=True, indexed=False)
+    # The response submitted by the learner.
+    learner_answer_details = ndb.StringProperty(required=True, indexed=False)
+    # The time at which the response was created.
+    created_on = ndb.DateTimeProperty(required=True, indexed=True)
+    # Whether the response received has been resolved.
+    is_resolved = ndb.BooleanProperty(default=False, required=True)
+
+    @classmethod
+    def get_entity_id(cls, id_parameters, entity_type):
+        """Gets entity_id for a batch model based on given entity_type.
+
+        Args:
+            id_parameters: dict. The parameters which will be
+                responsible to generate the entity_id.
+            entity_type: str. The type of the entity.
+
+        Returns:
+            str. Returns entity_id for a new instance of this class.
+
+        Raises:
+           Exception: If entity type is not valid.
+        """
+        # NOTE: In future if this model is used for any other entity type,
+        # then make the id_parameters dict for it, for eg. id_parameters
+        # dict for entity type exploration is {'exp_id': '', 'state_name': ''}
+        #  and add an elif condition in this function with the method to
+        #  generate the entity_id for that entity.
+        if entity_type == feconf.ENTITY_TYPE_EXPLORATION:
+            return '%s.%s' % (
+                id_parameters['exp_id'], id_parameters['state_name'])
+        else:
+            raise Exception('Invalid entity type.')
+
+    @classmethod
+    def generate_new_answer_details_id(cls, entity_type, entity_id):
+        """Generates a new answer detail ID which is unique.
+
+        Args:
+            entity_type: str. The type of the entity.
+            entity_id: str. The ID of the entity.
+
+        Returns:
+            str. An answer detail ID that is different from the IDs of all
+                the existing answer details within the given entity.
+
+        Raises:
+           Exception: There were too many collisions with existing answer
+               detail IDs when attempting to generate a new answer detail ID.
+        """
+        for _ in range(_MAX_RETRIES):
+            answer_details_id = (
+                entity_type + '.' + entity_id + '.' +
+                utils.base64_from_int(utils.get_current_time_in_millisecs()) +
+                utils.base64_from_int(utils.get_random_int(_RAND_RANGE)))
+            if not cls.get_by_id(answer_details_id):
+                return answer_details_id
+        raise Exception(
```

Comment:
Also @aks681, only this line is remaining, as it can't be covered.

Reply:
Ah, ok. This line is an extreme case anyway. It's fine then.

Reply:
It can be covered; talk to @lilithxxx for techniques (or see his recent PRs).